R/run-nmtran.R

```diff
@@ -0,0 +1,330 @@
+    .nonmem_version = NULL,
+    .config_path = NULL,
+    run_dir = tempdir(),
+    clean = TRUE
+){
+  check_model_object(.mod, "bbi_nonmem_model")
+
+  # Capture NONMEM/NM-TRAN options and format `NM-TRAN` args
+  nmtran_specs <- nmtran_setup(.mod, .nonmem_version, .bbi_args, .config_path)
+
+  mod_path <- get_model_path(.mod)
+  data_path <- get_data_path_from_ctl(.mod)
+
+  # Make temporary directory in current directory
+  mod_name <- fs::path_ext_remove(basename(mod_path))
+  temp_folder <- withr::local_tempdir(
+    pattern = paste0("nmtran-mod_", mod_name, "-"),
+    tmpdir = run_dir, clean = clean
+  )
+
+  # Copy model
+  file.copy(mod_path, temp_folder)
+  nmtran_mod <- new_model(file.path(temp_folder, basename(mod_path)))
+
+  # Copy dataset & overwrite $DATA record of new model
+  # NM-TRAN will error if data cannot be found
+  if(fs::file_exists(data_path)){
+    file.copy(data_path, temp_folder)
+    # overwrite $DATA record of new model
+    modify_data_path_ctl(nmtran_mod, data_path = basename(data_path))
+  }
+
+  # Run NM-TRAN
+  nmtran_results <- c(
+    list(
+      nmtran_exe = nmtran_specs$nmtran_exe,
+      nmtran_presort_exe = nmtran_specs$nmtran_presort_exe,
+      nonmem_version = nmtran_specs$nonmem_version,
+      absolute_model_path = mod_path
+    ),
+    execute_nmtran(
+      nmtran_specs$nmtran_exe,
+      mod_path = basename(mod_path),
+      cmd_args = nmtran_specs$cmd_args,
+      nmtran_presort_exe = nmtran_specs$nmtran_presort_exe,
+      dir = temp_folder
+    )
+  )
+
+  # assign class and return
+  class(nmtran_results) <- c(NMTRAN_PROCESS_CLASS, class(nmtran_results))
+  return(nmtran_results)
+}
+
+
+#' Sets up `NM-TRAN` to run by identifying an `NM-TRAN` executable for a
+#'  provided `NONMEM` version and formatting arguments based on `.bbi_args`,
+#'  the corresponding `.mod` yaml file, and any `nmfe_options` defined in a
+#'  `bbi.yaml` file
+#'
+#' @inheritParams run_nmtran
+#' @keywords internal
+nmtran_setup <- function(
+    .mod = NULL,
+    .nonmem_version = NULL,
+    .bbi_args = NULL,
+    .config_path = NULL
+){
+  bbi_yaml_path <- get_bbi_yaml_path(.mod, .config_path = .config_path)
+  bbi_yaml <- yaml::read_yaml(bbi_yaml_path)
+
+  # Combine NONMEM submission args
+  #  - The main ones of interest are prdefault, tprdefault, and maxlim, which
+  #    impact the evaluation of `NM-TRAN`
+  #  - Priority: .bbi_args > model yaml > bbi.yaml
+  cmd_args <- parse_nmtran_args(
+    .mod, .bbi_args = .bbi_args, nmfe_options = bbi_yaml$nmfe_options
+  )
+
+  # Check nonmem version
+  nm_config <- bbi_yaml$nonmem
+  if (!is.null(.nonmem_version)) {
+    # check for valid version
+    if (!(.nonmem_version %in% names(nm_config))) {
+      rlang::abort(
+        c(
+          "Must specify a valid `.nonmem_version` for run_nmtran.",
+          "i" = glue("{bbi_yaml_path} contains the following options:"),
+          glue("`{paste(names(nm_config), collapse='`, `')}`")
```

Comment:
[ nit-pick and side remark, okay not to address ]

This would render an error like so:

```R
nmtran_setup(mod1, .nonmem_version = "foo" )
#> Error in `nmtran_setup()`:
#> ! Must specify a valid `.nonmem_version` for run_nmtran.
#> ℹ /data/home/kylem/src/github/metrumresearchgroup/bbr/inst/model/nonmem/basic/bbi.yaml contains the following options:
#> `nm73gf`, `nm73gf_nmfe`, `nm74gf`, `nm74gf_nmfe`, `nm75`
#> Run `rlang::last_trace()` to see where the error occurred.
```

The \`nm\*\` markup is confusing to me because these are string values for `.nonmem_version`. In my opinion it'd be best to just leave them bare, but quoting with regular or double quotes would be better than using backticks.

Also perhaps worth noting: now that the minimum cli version is 3.1.0, it'd be safe to use `cli_abort`, in which case you could do something like

```diff
diff --git a/R/run-nmtran.R b/R/run-nmtran.R
index 0cfee6c4..b920fbb5 100644
--- a/R/run-nmtran.R
+++ b/R/run-nmtran.R
@@ -129,11 +129,11 @@ nmtran_setup <- function(
   if (!is.null(.nonmem_version)) {
     # check for valid version
     if (!(.nonmem_version %in% names(nm_config))) {
-      rlang::abort(
+      cli::cli_abort(
         c(
-          "Must specify a valid `.nonmem_version` for run_nmtran.",
-          "i" = glue("{bbi_yaml_path} contains the following options:"),
-          glue("`{paste(names(nm_config), collapse='`, `')}`")
+          "Must specify a valid {.var .nonmem_version} for run_nmtran.",
+          "i" = "{.path {bbi_yaml_path}} contains the following options:",
+          "{names(nm_config)}"
         )
       )
     }
```

```R
nmtran_setup(mod1, .nonmem_version = "foo" )
#> Error in `nmtran_setup()`:
#> ! Must specify a valid `.nonmem_version` for run_nmtran.
#> ℹ '/data/home/kylem/src/github/metrumresearchgroup/bbr/inst/model/nonmem/basic/bbi.yaml' contains the following options:
#> nm73gf, nm73gf_nmfe, nm74gf, nm74gf_nmfe, and nm75
#> Run `rlang::last_trace()` to see where the error occurred.
```